### PR TITLE
Replace magic numbers with keycode constants #4929

### PR DESCRIPTION
**Pull request by callmetushar123**
Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: Iccb2e4e89381a78337a3123543d01b00f9f24e6f

* Resolves: # 
* Target version: master 

### Summary
Just added some of the keystrokes, much more needed to be done

### TODO

- [x] Need to completely map all the bindings

### Checklist

- [x] Code is properly formatted
- [x] All commits have Change-Id
- [ ] I have run tests with `make check`
- [ ] I have issued `make run` and manually verified that everything looks okay
- [x] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Comment by callmetushar123** [comment]
@pedropintosilva Hello, I have discarded the previous pull request, made a new one, please check if everything is alright.
A lot more is needed to be done before merging

**Review by eszkadev — CHANGES_REQUESTED** [review]
hi, what we want here is to replace numbers with constants in conditions like this one somewhere below:
`if (shift && (ev.keyCode === 45 || ev.keyCode === 46)) {`

so we will replace 45 with human readable name :)

**Comment by callmetushar123** [comment]
Hello @eszkadev what I understood is that i need to create a mapping from keys to their integer values in keyCodes map, and then replace the corresponding number to their names. Am i right?

PS: can you please put the assign label on me?

**Comment by callmetushar123** [comment]
I have two doubts, in the keymap, keys from `16` to `20` has null as it's in corresponding values. Why? Am I supposed to assign the commented keyboard name to it?
Secondly `NUM0` has two keybindings `48` and `96`, should I follow the same in my keyCode map?

**Comment by eszkadev** [comment]
> I have two doubts, in the keymap, keys from `16` to `20` have null as it's in corresponding values.

Sorry, I don't know what you are talking about.
Where do you have these values? could you provide a link?
What does mean 'key 16'?

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [inline browser/src/map/handler/Map.Keyboard.js:150]: @eszkadev I was talking about this 16, Is this okay to put it like this?

Comment (eszkadev) [reply browser/src/map/handler/Map.Keyboard.js:150]: I see no problem with that (assuming value is correct) :)
But why should it be also UNKNOWN?

Comment (callmetushar123) [reply browser/src/map/handler/Map.Keyboard.js:150]: I got the value from `16  : null, // shift` which is present in the `keymap` dictionary of the same file

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [reply browser/src/map/handler/Map.Keyboard.js]: Okay, In `keymap` 13 had `return`, I will change it now.

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [inline browser/src/map/handler/Map.Keyboard.js:706]: @eszkadev I am stuck here, according to the comments, it wants a small 'c' and 'x' but there is no dictionary key-value pairs to replace them, So I am forced to write `NUM3 ` whose corresponding value is 120.
Any leads on what to do?

Comment (eszkadev) [reply browser/src/map/handler/Map.Keyboard.js:706]: you can create multiple labels with the same value there...
keyCodes.c = 99
keyCodes.NUM3 = 99

but this makes me suspicious, maybe you mixed ASCII codes with some KEY CODES for events here?
Because it should be not ambigous, key codes probably have only one 'C' letter (no small and big variant).
For example if we want big characters - we use the same character code but combined (added) with SHIFT code.

File: browser/src/map/handler/Map.Keyboard.js
Comment (eszkadev) [reply browser/src/map/handler/Map.Keyboard.js:150]: 'shift: unknown' comment doesn't appy to this enum, it was about UNO codes - it's something different.

Comment (eszkadev) [reply browser/src/map/handler/Map.Keyboard.js:150]: please do not look at UNO key codes, that is something completly different, only for our server. here we want to make names for JS key codes.

File: browser/src/map/handler/Map.Keyboard.js
Comment (eszkadev) [reply browser/src/map/handler/Map.Keyboard.js:706]: As you can see in the code we compare values of `keyCodes` map with `ev.keyCode`.
Documentation is here: https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode

Comment (callmetushar123) [reply browser/src/map/handler/Map.Keyboard.js:706]: > Because it should be not ambigous, key codes probably have only one 'C' letter (no small and big variant).

But this raises another question, if There is no small and big variant of 'C' then shouldn't the hardcoded value be `67 `instead of `99`?

Comment (eszkadev) [reply browser/src/map/handler/Map.Keyboard.js:706]: I think we don't need all the possible values in the map. Only used in the code are enough :)
So from ASCII characters we need only 'c', 'x', 'v', 'z', 'p', 's' if I see correctly :)
Also probabyl we don't need these arrays for NUMX labels...

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [inline browser/src/map/handler/Map.Keyboard.js:164]: There were two values for `NUM0 `in `keymap`. I have clubbed them together like this in `keyCodes`, Is my approach right?

Comment (eszkadev) [reply browser/src/map/handler/Map.Keyboard.js:164]: one for safari/iOS, other for other systems? 

If it's needed and true lets keep it that way :)

File: browser/src/map/handler/Map.Keyboard.js
Comment (eszkadev) [reply browser/src/map/handler/Map.Keyboard.js:706]: > > Because it should be not ambiguous, key codes probably have only one 'C' letter (no small and big variant).
> 
> But this raises another question, if There is no small and big variant of 'C' then shouldn't the hardcoded value be `67 `instead of `99` in line 706?

documentation says we shouldn't use that field of an event for ASCII characters at all (should be code instead of keyCode - but it's not our problem here)... so... let's keep the code unchanged - we only want human readable names so it's easier to read :)

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [reply browser/src/map/handler/Map.Keyboard.js:164]: > one for safari/iOS, other for other systems?

yes. Okay will keep it

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [reply browser/src/map/handler/Map.Keyboard.js:706]: >I think we don't need all the possible values in the map. Only used in the code are enough :)
So should I go with this solution?
keyCodes.c = 99
keyCodes.NUM3 = 99

**Comment by callmetushar123** [comment]
So I figured out that for windows the value for A will remain the same if we perform a `keydown `or not, but for safari, a `keydown `change the value of A, So I needed to make a separate mapping accordingly. 

Ref: [link](https://css-tricks.com/snippets/javascript/javascript-keycodes/)

@eszkadev please check and approve the merge

File: browser/src/map/handler/Map.Keyboard.js
Comment (eszkadev) [inline browser/src/map/handler/Map.Keyboard.js:145]: Can we move enter after `pageDown:` line as it was before so it will not show modification?
Later it's easier to find last commit using git grep tool.

Comment (eszkadev) [reply browser/src/map/handler/Map.Keyboard.js:145]: (We will fix order and name later in separate PR, just want to keep access to the commit introducing that variable :) )

File: browser/src/map/handler/Map.Keyboard.js
Comment (eszkadev) [inline browser/src/map/handler/Map.Keyboard.js:531]: 36 should be HOME
(LEFT is 37)

File: browser/src/map/handler/Map.Keyboard.js
Comment (eszkadev) [inline browser/src/map/handler/Map.Keyboard.js:585]: hmm, it shouldn't be misleading:
`keyCodes.NUM3[1]` -> `keyCodes.c`  - as we already have dedicated name for it

File: browser/src/map/handler/Map.Keyboard.js
Comment (eszkadev) [inline browser/src/map/handler/Map.Keyboard.js:591]: `keyCodes.NUM3[1]` -> `keyCodes.c`

File: browser/src/map/handler/Map.Keyboard.js
Comment (eszkadev) [inline browser/src/map/handler/Map.Keyboard.js:198]: I think we should unify that with the NUMX approach -> use an array with 2 values here V: [86, 118]

We can also add name for accessing those:
keyCodes.NUM3[DEFAULT] or keyCodes.NUM3[MAC]
where DEFAULT = 0, MAC = 1

**Review by eszkadev — CHANGES_REQUESTED** [review]
Nice work! just few details to polish that and we can merge :)

File: browser/src/map/handler/Map.Keyboard.js
Comment (eszkadev) [reply browser/src/map/handler/Map.Keyboard.js]: we can unify approach, use arrays for all the cases:
NUM3: [ _, _ ]
C: [ _, _ ]

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [reply browser/src/map/handler/Map.Keyboard.js:585]: How did you know `99` here refers to `c`? Was it by looking at the if statement above it? If Yes then I guess I need to change `118` and `120` too.

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [reply browser/src/map/handler/Map.Keyboard.js]: Done

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [reply browser/src/map/handler/Map.Keyboard.js:145]: Done

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [reply browser/src/map/handler/Map.Keyboard.js]: Done, all newly added were upper cased

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [reply browser/src/map/handler/Map.Keyboard.js]: Done

File: browser/src/map/handler/Map.Keyboard.js
Comment (eszkadev) [inline browser/src/map/handler/Map.Keyboard.js:719]: it was ` 93: // Right Cmd (Safari)`
but `RIGHTWINDOWKEY` is  `92`

File: browser/src/map/handler/Map.Keyboard.js
Comment (eszkadev) [inline browser/src/map/handler/Map.Keyboard.js:665]: C is an Array now
`...C[DEFAULT]` here

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [reply browser/src/map/handler/Map.Keyboard.js:719]: Oh I missed it. Will do it.

**Review by eszkadev — CHANGES_REQUESTED** [review]
I found 2 small mistakes

**Comment by callmetushar123** [comment]
Okay I will fix them ASAP

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [reply browser/src/map/handler/Map.Keyboard.js]: done

File: browser/src/map/handler/Map.Keyboard.js
Comment (callmetushar123) [reply browser/src/map/handler/Map.Keyboard.js]: done

**Comment by eszkadev** [comment]
Now it looks ok :) Currently we cannot merge any risky changes so we have to wait a bit (week I think).

Could you also squash your commits into one? It will be easier to read the change later.

**Comment by callmetushar123** [comment]
The last time I squashed I had problems like you see the first pick in the picture, It's not my commit but it came somehow in the list of commits to be squashed. Do you know why it happened?
squash

Also, should I sync my forked repo?

**Comment by callmetushar123** [comment]
Squashed all my commits together

**Comment by pedropintosilva** [comment]
> Squashed all my commits together

Awesome, Thanks! Could you please amend your commit with `git commit --amend` and fix the commit subject AKA title (first line) and the commit description? Right now your commit message looks like:

```
First commit just made some few changes

and also squashed all my commits together
Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: Iccb2e4e89381a78337a3123543d01b00f9f24e6f

Made all the required changes

Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: Ifea5cf1a8bf73fc91a8b541104d23b5e5ec46981

changed keyCodes to keyCode

Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: I40be65fa5c567ee2f722d78f6c226444546cb207

changed keyCode to this.keyCodes

Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: I257a55fbd6a64f4e83f04d22892acbc53bca8aae

solved duplicate keys issue

Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: I5de1e893996ffb626c055295bbc87a8f76cce624

changed return to enter in keyCodes dictionary

Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: I54c44915c0ebd5b19f82bc699010708b411f3622

line 343 had errors, resolved it

Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: I29e1556ba9e31fa74bfdb5d84c1ad06fbc39e396

duplicate subtract error, resolved it

Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: Ic2b301f28cc9a0ed82df0fee3b215cc77530c86d

line 698, changed num5 to num5[0]

Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: Id9aed56375e9d42911ec40267f3662bc783d86be

resolved conflicts related to keymappings

Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: I8ddacf568eb560d2992b662c96a6a32825726700

 Made capital letters and few other changes

Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: I477bf4d9b2fb469d7173937ca39b3a7e85f4e118

 Fixed MAC value not used

Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: I654854f224ccfef0b5a477c184fa5a797c6a4795

 Fixed left and right window key for MAC
Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: I1d3631099454c4846352fd475307f03f33f2038a
```

Please do a `git commit --amend` (the text editor will open) and rewrite it so it looks something like:

```
Replace magic numbers with keycode constants

There are directly provided numbers for key codes in some places. We can use already defined constant variables. This makes code more readable and obvious. This fixes issue #4929 .

Signed-off-by: Tushar <tusharbarman123@gmail.com>
Change-Id: Iccb2e4e89381a78337a3123543d01b00f9f24e6f
```

**Comment by callmetushar123** [comment]
@pedropintosilva Do I need to perform git rebase -i HEAD~1 or I can directly amend and push?

**Comment by callmetushar123** [comment]
> @pedropintosilva Do I need to perform `git rebase -i HEAD~1` or I can directly amend and push?

**Comment by pedropintosilva** [comment]
> @pedropintosilva Do I need to perform git rebase -i HEAD~1 or I can directly amend and push?

You can directly amend it and then push : adding additional info in my comment ^ above

**Comment by eszkadev** [comment]
> Can I sync my forked repo as I have to make another pull request in the same project? It will not affect this pull request, right?

You need to use different branch for every PR, but repo can be the same.

**Review by eszkadev — APPROVED** [review]
looks ok now, thanks :)

**Comment by pedropintosilva** [comment]
Thanks @eszkadev for reviewing and thank you very much and welcome aboard @callmetushar123  :partying_face:

**Pull request merged**
